[step-3] 지하철 구간

heozeop opened the pull request:
## 작업
### 인수 테스트 작성
- 조건에 대한 인수 테스트 작성

### 서비스 코드 작성
- 구간 추가 / 삭제

heozeop committed:
refac: rename lines to line

heozeop committed:
test: delete outdated dirties context

heozeop committed:
test: use recursive comparison on comparing lines

heozeop committed:
test: add check on null value for get item

heozeop committed:
docs: add requirements

heozeop committed:
test: add acceptance test structure for section

heozeop committed:
test: setup initialize and database clean up

heozeop committed:
docs: add request data

heozeop committed:
test: add test for add section

heozeop committed:
feat: add sections instead of direct link for station

heozeop committed:
feat: add section api on line controller

heozeop committed:
test: fix section add test

heozeop committed:
test: add test for down station validation

heozeop committed:
feat: add validation for add section

heozeop committed:
refac: rename lines to line of service function name

heozeop committed:
test: add delete section test

heozeop committed:
feat: add delete section api

heozeop committed:
test: add delete section validation test for no extra section exist

heozeop committed:
feat: add validate section number on delete

heozeop committed:
test: add test for delete not the end point

heozeop committed:
feat: add validation condition on delete and make up line entity status

heozeop committed:
docs: update requirements

heozeop commented on src/test/java/subway/SubwaySectionAcceptanceTest.java:
```diff
@@ -0,0 +1,198 @@
+        // Then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.NO_CONTENT.value());
+
+        // Then
+        final List<Long> stationIdList = lineService
+            .getLine(lineId)
+            .getStations().stream().map(StationResponse::getId)
+            .toList();
```
1. lineService를 이용해 데이터를 조회하는 것에 대해 어떻게 생각하시나요?
    1. RestAssured를 이용해 자원을 가져오는 패턴이 더 좋은 패턴이라고 생각하시는지 궁금합니다.

pch8388 replied:
`인수테스트` 는 시나리오에 대해 테스트 하기 때문에 가능하면 내부적 의존성을 갖지 않는 게 좋다고 생각합니다 😄 
지금 프로젝트와 같은 상황에서는 api 호출만 사용한다면, 내부 구현에 상관없이 시나리오를 확인해볼 수 있습니다.
`LineService`가 만약 지하철 역들을 반환하지 않고, `StationService` 가 반환하도록 변경한다면 어떻게 될까요? 🤔

heozeop commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,157 @@
+            )
+        );
+        section.updateLine(line);
+        line.updatesSection(section.getDownStationId(), line.getDistance() + section.getDistance());
```
1. line에 저장된 최종역 등을 갱신하는 요구사항에 대한 테스트는 어떻게 진행하는게 좋을까요?

pch8388 replied:
`line.updateSection`의 기능 자체는 `LineTest`에서 단위테스트를 만들면 될 것 같습니다.
그 이상의 단위를 테스트 할때는 그에 맞게 해주시면 됩니다. 😄 (ServiceTest 나 인수테스트에서는 갱신된 정보를 조회해서 원하는 대로 변경되었는지 확인한다던가 하는 식으로) 
사실 테스트를 얼마나 만들어야 하나에 대한 정답은 없어서, 시간이 가능한 충분히 만들어야 한다..라는 애매한 답변을 드릴수 밖에 없네요.. 😢

heozeop commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,62 @@
+        return distance;
+    }
+
+    public void updateLine(Line line) {
+        this.line = line;
+        line.addSection(this);
+    }
```
1. 매핑을 양방향으로 해주었는데, 그냥 생성자에서 넣어주는 방식이 더 좋은 방식일까요?
    1. 반대쪽에서 조회할때 성능에 차이가 있을까 싶어 적용해 보았습니다.

pch8388 replied:
실제로 발생하는 쿼리를 보시면 좋을 것 같아요 😅 
대부분의 방법은 트레이드오프가 있습니다. 성능 문제는 항상 기반 상황(인프라 등)이나 데이터 모델, 데이터 양 등 여러가지 요소를 고려해서 확인해야 하기 때문에, 신중하게 최적화 해야합니다
현재 강의에서 추구하는 모델은 하나의 구간은 하나의 노선만 가지기 때문에, 생성자에서 넣어주시는 것이 더 논리적으로 맞다고 생각이 듭니다 🤔 
우려하시는 조회시의 성능에 대해서는 조금 더 자세히 설명해주시면 답변 드릴 수 있을 것 같아요

heozeop reviewed (commented):
리뷰 요청드립니다!

pch8388 commented on src/main/java/subway/lines/Line.java:
```diff
@@ -0,0 +1,79 @@
+        this.color = color;
+    }
+
+    public void updatesSection(Long downStationId, Long distance) {
```
`updatesSection` 이라는 네이밍이랑은 다르게 downStation 에 대한 정보를 수정하는 것으로 보여요 👀

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,157 @@
+    public LineResponse saveLine(LineCreateRequest lineCreateRequest) {
+        final Line line = lineRepository.save(
+            new Line(
+                lineCreateRequest.getName(),
```
lineCreateRequest 내부에서 Line 을 생성하는 메소드를 만들어도 좋을 것 같고, lineCreateRequest -> line 으로 변환해주는 Factory 성격의 클래스를 만들어도 좋을 것 같습니다. 🤔 
객체 생성에 많은 라인이 할당되어 비즈니스에 집중이 안될 수 있을 것 같아요 😅

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,157 @@
+    public LineResponse getLine(Long id) {
+        final Line line = lineRepository.findById(id).orElse(null);
+        if (line == null) {
+            return null;
```
[코멘트](https://github.com/next-step/atdd-subway-map/pull/1025#discussion_r1468719145) 에서 말씀드린 것처럼 null 을 직접 반환하는 것이 괜찮은지 생각해보시면 좋을 것 같아요 🤔

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,157 @@
+
+    @Transactional
+    public void updateLines(Long id, LineUpdateRequest lineUpdateRequest) {
+        final Line line = lineRepository.findById(id).orElse(null);
```
[코멘트](https://github.com/next-step/atdd-subway-map/pull/1025#discussion_r1468719872) 참고해주세요 😢

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,157 @@
+            throw new IllegalArgumentException();
+        }
+
+        if(line.getDownStationId() != stationId) {
```
wrapper type 비교시에는 equals 로 비교 하셔야 합니다 😄 
`Long.valueOf(200) == Long.valueOf(200);`  는 false 입니다

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    private void validateDeleteSectionCondition(Long id, Long stationId, Line line) {
+        final int numberOfSection = sectionRepository.countByLineId(id);
```
비즈니스 로직이 도메인 객체에 존재하도록 구현해보세요 😅

pch8388 commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,62 @@
+package subway.section;
+
+import javax.persistence.CascadeType;
```
미사용 import 는 삭제해주세요 👀

pch8388 commented on src/test/java/subway/SubwaySectionAcceptanceTest.java:
```diff
@@ -0,0 +1,198 @@
+
+    @BeforeEach
+    void setUp() {
+        upStationId = stationRepository.save(new Station("역1")).getId();
```
여기도 Repository 를 직접 사용하는 것보다 `StationAcceptanceTest` 에서 사용한 메소드를 사용해주세요 😸 
인수테스트는 정의한 시나리오가 전부 정상적으로 작동되는 지를 확인하는 것이기 때문에 실제 클라이언트의 흐름과 최대한 동일하게 만들어야 합니다

pch8388 commented on src/main/java/subway/lines/LinesController.java:
```diff
@@ -3,57 +3,85 @@
+        SectionAddRequest sectionAddRequest) {
+        try {
+            return ResponseEntity.ok().body(lineService.addSection(id, sectionAddRequest));
+        } catch (EntityNotFoundException e) {
```
[예외처리](https://cheese10yun.github.io/spring-guide-exception/) 참고해보시면 어떨까요? 🤔 
발생하는 처리해야할 예외 종류가 늘어날수록 코드 복잡성이 올라갈 것 같아요 😢

pch8388 commented on src/test/java/subway/SubwaySectionAcceptanceTest.java:
```diff
@@ -0,0 +1,198 @@
+            downStationId,
+            extraStationId,
+            10L
+        );
```
이 구문 아래에 다음 구문을 추가해도 정상 동작합니다 😢 

```java
final ExtractableResponse<Response> response2 = addSections(
                lineId,
                extraStationId,
                downStationId,
                10L
        );
```

결과는 이렇게 만들어 집니다
역1 - 역2
역2 - 역3
역3 - 역2

pch8388 reviewed (changes requested):
3단계 미션 잘수행해주셨는데, 몇가지 수정하였으면 하는 부분이 있습니다 😄 
코멘트 확인하시고 수정 후 리뷰요청 해주세요 🔥

heozeop committed:
chore: delete unused import

heozeop committed:
refac: make validate inside of line domain

heozeop committed:
test: split access test with api

heozeop committed:
feat: add section on line to update section

heozeop committed:
test: split line api request from test

heozeop committed:
refac: add final on property

heozeop committed:
test: handle redundant station add error case

heozeop committed:
refac: throw on not found

heozeop committed:
refac: throw on not found on get line

heozeop committed:
refac: make final for all repository on line service

heozeop committed:
refac: update line on get section

heozeop committed:
refac: use global exception handler

heozeop committed:
chore: delete unused import

heozeop committed:
docs: add prompt for new error case

heozeop committed:
chore: add space on eof

heozeop committed:
refac: delete not used property

heozeop committed:
feat: make global error handler work

heozeop commented:
스프링의 예외 처리를 공부하려고 합니다. 리뷰어님은 어떻게 공부하셨는지 궁금합니다!

pch8388 commented:
> 스프링의 예외 처리를 공부하려고 합니다. 리뷰어님은 어떻게 공부하셨는지 궁금합니다!

[커스텀 예외](https://tecoble.techcourse.co.kr/post/2020-08-17-custom-exception) 참고해보시면 좋아요 😄 

앗 이미 링크를 드렸었군요.. 🤔 
예외처리 부분도 경험적인 부분이 커서 링크의 글 참고하시고 해보면서 좋은 방법을 찾는 것이 좋습니다

heozeop committed:
feat: add missing file

pch8388 commented on src/main/java/subway/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
💯

pch8388 commented on src/main/java/subway/section/SectionAddRequest.java:
```diff
@@ -23,4 +25,11 @@ public Long getDownStationId() {
         return distance;
     }
+
+    public Section getSection(Line line) {
```
지금 이 클래스는 [DTO](https://martinfowler.com/eaaCatalog/dataTransferObject.html) 성격의 클래스라고 할 수 있는데요
도메인 클래스의 메소드를 호출하는 것과 같은 비즈니스에 관련된 일은 하지 않는 것이 좋습니다 😢

pch8388 commented on src/main/java/subway/lines/Line.java:
```diff
@@ -0,0 +1,108 @@
+    private Long downStationId;
+    private Long distance;
+
+    @OneToMany(mappedBy = "line")
```
```suggestion
    @OneToMany(mappedBy = "line", cascade = {CascadeType.PERSIST, CascadeType.MERGE}, orphanRemoval = true)
```

이런형태로 해도 될 것 같아요 😄 
`구간은 노선에 포함되어있다` 라는 전제를 가져도 될만한 상황이기 때문에, 라이프 사이클을 공유하도록 구현하셔도 좋습니다.
위 옵션들에 대해서는 [참고](https://velog.io/@yuseogi0218/JPA-CascadeType.REMOVE-vs-orphanRemoval-true) 를 봐주세요 

이렇게 해두면 조금 더 객체지향적인 설계를 하실 수 있을 거에요
일단은 한번 도전해보시고, 너무 오래 걸리거나 힘드시면 따로 말씀해주세요 😅 
참고로 말씀드리면 목요일 강의전에 해당 주차의 미션을 무조건 다 완수하지 않으셔도 됩니다.

pch8388 commented on src/main/java/subway/lines/Line.java:
```diff
@@ -0,0 +1,108 @@
+    private Long distance;
+
+    @OneToMany(mappedBy = "line")
+    private List<Section> sections = new ArrayList<>();
```
[일급컬렉션](https://jojoldu.tistory.com/412)도 가능하면 도전해보세요 🔥

pch8388 commented on src/main/java/subway/lines/Line.java:
```diff
@@ -0,0 +1,108 @@
+        this.color = color;
+    }
+
+    public void updateSectionInfo(Long downStationId, Long distance) {
```
상황과 조금 어울리지 않는 것 같아요 😢 
만약 구간이 삭제된 후에 sections 로 무언가를 할일이 있다면, 지금 이것이 제대로 된 구현일까요? 🤔

pch8388 commented on src/main/java/subway/lines/Line.java:
```diff
@@ -0,0 +1,108 @@
+        return sections;
+    }
+
+    public void validateSectionToAdd(SectionAddRequest sectionAddRequest) {
```
전체적으로 도메인 객체에서 외부 의존성(DTO 등)을 갖는 것을 제거해주세요

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,113 @@
+            lineCreateRequest.getDownStationId(),
+            lineCreateRequest.getDistance()
+        );
+        section.updateLine(line);
```
현재 도메인 정책상 `구간`은 항상 `노선`에 포함될 것 같아서, 생성자에서 line 을 할당해도 될 것 같아요 😄

pch8388 commented on src/test/java/subway/StationApiRequester.java:
```diff
@@ -0,0 +1,72 @@
+import org.springframework.http.MediaType;
+import subway.station.StationResponse;
+
+public final class StationApiRequester {
```
잘 구현해 주셨네요 👍

pch8388 commented on src/test/java/subway/StationApiRequester.java:
```diff
@@ -0,0 +1,72 @@
+        RestAssured.given().log().all()
+            .body(params)
+            .contentType(MediaType.APPLICATION_JSON_VALUE)
+            .when().put("/stations/" + id)
```
```suggestion
            .when().put("/stations/{stationId}", id)
```

위와 같은 형태로 하면 가독성이 더 좋은 것 같아요 😄

pch8388 reviewed (changes requested):
리팩토링 잘해주셨습니다 👏 
조금 더 고쳐보았으면 합니다 😅 
코멘트 참고하셔서 수정 후 리뷰요청 해주세요 🔥 🔥

heozeop committed:
feat: make domain object not to rely on dto

1. make cascade
1. rename update section info to delete section
1. make add section to work with valid add section

heozeop committed:
feat: make validation on dto

heozeop committed:
feat: make section to update line in constructor

heozeop committed:
feat: make remove section work

heozeop committed:
feat: update service to adapt changes

1. validate via dto
1. make add section and delete section work on line domain

heozeop committed:
feat: make section list to first class object

heozeop committed:
refac: handle add section on line

heozeop commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,106 @@
+        final Line line = lineRepository.findById(id).orElseThrow(EntityNotFoundException::new);
+        sectionDeleteRequest.validateToDelete(line);
+
+        final Section sectionToDelete = sectionRepository.findByLineIdAndDownStationId(id, sectionDeleteRequest.getStationId()).orElseThrow(EntityNotFoundException::new);
```
line의 getSections 함수를 이용해서 찾는 방식이 객체지향적으로 더 좋은 방식일까요?
validation 할 때 getSections함수를 호출하여 select 쿼리가 나가는 관계로 쿼리를 한번 더 던지는게 옳을까 싶어 문의드립니다.

pch8388 replied:
객체지향적으로 좋은 지에 대해서 너무 깊게 고민하시는 것보다 설계를 어떤 관점에서 일관성 있게 가져갈 것이가에 대해 고민해보시면 좋습니다. 😄 (이런 이야기는 다른 주차 리뷰어에게 같은 질문을 해봐도 좋을 것 같아요. 리뷰어마다도 의견이 다를 수 있는 문제라고 생각합니다)

제 생각을 정리해보겠습니다.
지금 도메인에서는 일단은 `구간`이 외부로 드러나지 않도록 구현을 해볼 것 같아요  🤔 
`구간`이 외부로 드러나지 않게 하려다보면 `line` 내부에서 `삭제할 구간을 찾아 제외한다` 라는 로직을 처리할 수 밖에 없을 거에요
그렇게 처리하면 Line 외부의 관점에서 메소드를 사용하는 것이 다음과 같이 될 거라고 예상됩니다.

```java
line.deleteSection(targetStationId);
```

pch8388 commented on src/main/java/subway/section/SectionAddRequest.java:
```diff
@@ -27,9 +30,22 @@ public Long getDistance() {
+        return new Section(line, upStationId, downStationId, distance);
+    }
+
+    public void validateSectionToAdd(Line line) {
```
DTO 클래스에 비즈니스 로직이 담기게 됩니다 😢 
line 이나 section 이 판단할 수 있도록 해주세요

pch8388 commented on src/main/java/subway/section/SectionDeleteRequest.java:
```diff
@@ -10,4 +14,15 @@ public SectionDeleteRequest(Long stationId) {
         return stationId;
     }
+
+    public void validateToDelete(Line line) {
```
여기도 DTO 에서 비즈니스 validation 을 판단하지 않도록 해주세요 😭

pch8388 commented on src/main/java/subway/lines/Line.java:
```diff
@@ -27,8 +22,8 @@ public class Line {
-    @OneToMany(mappedBy = "line")
-    private List<Section> sections = new ArrayList<>();
+    @Embedded
+    private SectionList sectionList = new SectionList();
```
`SectionList` 는 특정 자료구조에 의존적인 네이밍인것 같습니다 🤔 
복수형을 나타낼 수 있도록만 네이밍 하는 것이 더 좋습니다 😸

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -32,14 +32,14 @@ public LineService(LineRepository lineRepository, StationRepository stationRepos
-            lineCreateRequest.getDownStationId(),
-            lineCreateRequest.getDistance()
+        line.addSection(
+            new Section(
```
외부에서 `Section` 을 생성해야할만한 이유가 있는 지도 고민해보시면 좋을 것 같아요 😄 
upStationId / downStationId / distance 를 넘기고 Line 에게 생성 책임을 넘겨도 될 거라고 생각됩니다
외부에 드러난 객체는 수정이 어렵습니다. 따라서, 가능하면 외부로 드러나는 객체를 한정 짓는 것이 좋습니다

pch8388 reviewed (changes requested):
리팩토링 잘해주셨는데, 조금만 더 해보면 좋을 것 같아요 😄 
코멘트 참고하셔서 수정 후 리뷰요청 해주세요 🔥

heozeop committed:
refac: rename sections and locate validation order on domain

1. sections: sectionList -> sections
1. validation in line domain: line need to handle add condition
1. make line to create section: separate service responsibility

heozeop committed:
test: fix test to pass with parameter

1. check down station on 지하철구간_종점제거_하행종점역이_아닌_경우: I thought that it is not a chance to test up station deletion

heozeop committed:
refac: delete dependency not in use

pch8388 commented on src/main/java/subway/section/SectionAddRequest.java:
```diff
@@ -0,0 +1,35 @@
+package subway.section;
+
+import java.util.HashSet;
```
미사용 import 👀

pch8388 commented on src/main/java/subway/section/SectionAddRequest.java:
```diff
@@ -0,0 +1,35 @@
+        return distance;
+    }
+
+    public Section getSection(Line line) {
```
이제 사용하지 않는 것 같은데 삭제해주세요 😄

pch8388 commented on src/main/java/subway/lines/Line.java:
```diff
@@ -0,0 +1,116 @@
+        this.distance += distance;
+    }
+
+    private void validateToAdd(Section sectionToAdd) {
```
이 메소드 자체를 `Sections`가 가져도 좋을 것 같아요 😄

pch8388 commented on src/main/java/subway/lines/LineService.java:
```diff
@@ -0,0 +1,103 @@
+    @Transactional
+    public void deleteSection(Long id, SectionDeleteRequest sectionDeleteRequest) {
+        final Line line = lineRepository.findById(id).orElseThrow(EntityNotFoundException::new);
+        final Section sectionToDelete = line.getSections().stream()
+            .filter(section ->
+                Objects.equals(section.getDownStationId(), sectionDeleteRequest.getStationId())
+            )
+            .findFirst()
+            .orElseThrow(EntityNotFoundException::new);
```
이 로직도 `Line`안으로 들어가도 됩니다 😅

pch8388 reviewed (approved):
리팩토링 잘해주셨습니다 👏 👏 
이번 미션은 여기서 마무리 하겠습니다.
남겨드린 코멘트는 확인만 해주셔도 될 것 같습니다 😄 
수고하셨습니다 🚀
